File: ios/Classes/Listeners/ConversationListener.swift
import Flutter
import TwilioConversationsClient

public class ConversationListener: NSObject, TCHConversationDelegate {
    let conversationSid: String
    
    init(_ conversationSid: String) {
        self.conversationSid = conversationSid
    }

    // onMessageAdded
    public func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation, messageAdded message: TCHMessage) {
        SwiftTwilioConversationsPlugin.debug(
            "ConversationListener.onMessageAdded => messageSid = \(String(describing: message.sid))")
        SwiftTwilioConversationsPlugin.flutterClientApi?.messageAddedConversationSid(
            conversationSid,
            messageData: Mapper.messageToPigeon(message, conversationSid: conversationSid),
            completion: { (_: Error?) in
                //TODO: consider logging an error
            })
    }

    // onMessageUpdated
    public func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation, message: TCHMessage, updated: TCHMessageUpdate) {
        SwiftTwilioConversationsPlugin.debug(
            "ConversationListener.onMessageUpdated => messageSid = \(String(describing: message.sid)), " +
                "updated = \(String(describing: updated))")
        SwiftTwilioConversationsPlugin.flutterClientApi?.messageUpdatedConversationSid(
            conversationSid,
            messageData: Mapper.messageToPigeon(message, conversationSid: conversationSid),
            reason: Mapper.messageUpdateToString(updated),
            completion: { (_: Error?) in
                //TODO: consider logging an error
            })
    }

    // onMessageDeleted
    public func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation, messageDeleted message: TCHMessage) {
        SwiftTwilioConversationsPlugin.debug(
            "ConversationListener.onMessageDeleted => messageSid = \(String(describing: message.sid))")
        SwiftTwilioConversationsPlugin.flutterClientApi?.messageDeletedConversationSid(
            conversationSid,
            messageData: Mapper.messageToPigeon(message, conversationSid: conversationSid),
            completion: { (_: Error?) in
                //TODO: consider logging an error
            })
    }

    // onParticipantAdded
    public func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation, participantJoined participant: TCHParticipant) {
        SwiftTwilioConversationsPlugin.debug(
            "ConversationListener.onParticipantAdded => participantSid = \(String(describing: participant.sid))")
        SwiftTwilioConversationsPlugin.flutterClientApi?.participantAddedConversationSid(
            conversationSid,
            participantData: Mapper.participantToPigeon(participant, conversationSid: conversationSid)!,
            completion: { (_: Error?) in
                //TODO: consider logging an error
            })
    }

    // onParticipantUpdated
    public func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation, participant: TCHParticipant, updated: TCHParticipantUpdate) {
        SwiftTwilioConversationsPlugin.debug(
            "ConversationListener.onParticipantUpdated => participantSid = \(String(describing: participant.sid)), " +
                "updated = \(String(describing: updated))")
        SwiftTwilioConversationsPlugin.flutterClientApi?.participantUpdatedConversationSid(
            conversationSid,
            participantData: Mapper.participantToPigeon(participant, conversationSid: conversationSid)!,
            reason: Mapper.participantUpdateToString(updated),
            completion: { (_: Error?) in
                //TODO: consider logging an error
            })
    }

    // onParticipantDeleted
    public func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation, participantLeft participant: TCHParticipant) {
        SwiftTwilioConversationsPlugin.debug(
            "ConversationListener.onParticipantDeleted => participantSid = \(String(describing: participant.sid))")
        SwiftTwilioConversationsPlugin.flutterClientApi?.participantDeletedConversationSid(
            conversationSid,
            participantData: Mapper.participantToPigeon(participant, conversationSid: conversationSid)!,
            completion: { (_: Error?) in
                //TODO: consider logging an error
            })
    }

    // onTypingStarted
    public func conversationsClient(_ client: TwilioConversationsClient, typingStartedOn conversation: TCHConversation, participant: TCHParticipant) {
        SwiftTwilioConversationsPlugin.debug(
            "ConversationListener.onTypingStarted => conversationSid = \(String(describing: conversation.sid)), " +
                "participantSid = \(String(describing: participant.sid))")
        SwiftTwilioConversationsPlugin.flutterClientApi?.typingStartedConversationSid(
            conversationSid,
            conversationData: Mapper.conversationToPigeon(conversation)!,
            participantData: Mapper.participantToPigeon(participant, conversationSid: conversationSid)!,
            completion: { (_: Error?) in
                //TODO: consider logging an error
            })
    }

    // onTypingEnded
    public func conversationsClient(_ client: TwilioConversationsClient, typingEndedOn conversation: TCHConversation, participant: TCHParticipant) {
        SwiftTwilioConversationsPlugin.debug(
            "ConversationListener.onTypingEnded => conversationSid = \(String(describing: conversation.sid)), " +
                "participantSid = \(String(describing: participant.sid))")
        SwiftTwilioConversationsPlugin.flutterClientApi?.typingEndedConversationSid(
            conversationSid,
            conversationData: Mapper.conversationToPigeon(conversation)!,
            participantData: Mapper.participantToPigeon(participant, conversationSid: conversationSid)!,
            completion: { (_: Error?) in
                //TODO: consider logging an error
            })

    }

    // onSynchronizationChanged
    public func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation, synchronizationStatusUpdated status: TCHConversationSynchronizationStatus) {
        SwiftTwilioConversationsPlugin.debug(
            "ConversationListener.onSynchronizationChanged => sid: \(String(describing: conversation.sid)), status: \(Mapper.conversationSynchronizationStatusToString(conversation.synchronizationStatus))")
        SwiftTwilioConversationsPlugin.flutterClientApi?.synchronizationChangedConversationSid(
            conversationSid,
            conversationData: Mapper.conversationToPigeon(conversation)!,
            completion: { (_: Error?) in
                //TODO: consider logging an error
            })
    }

    // The ConversationListener Protocol for iOS duplicates some of the events
    // that are provided via the ClientListener protocol on both Android and iOS.
    // In the interest of functional parity and avoid duplicate notifications,
    // we will not notify the dart layer of such event from the ConversationListener.
    public func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation, participant: TCHParticipant, userSubscribed user: TCHUser) {
        // userSubscribed
    }

    public func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation, participant: TCHParticipant, userUnsubscribed user: TCHUser) {
        // userUnsubscribed
    }

    public func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation, participant: TCHParticipant, user: TCHUser, updated: TCHUserUpdate) {
        // userUpdated
    }

    public func conversationsClient(_ client: TwilioConversationsClient, conversationDeleted conversation: TCHConversation) {
        // onConversationDeleted
    }

    public func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation, updated: TCHConversationUpdate) {
        // onConversationUpdated
    }
}

File: android/src/main/kotlin/twilio/flutter/twilio_conversations/listeners/ConversationListener.kt
package twilio.flutter.twilio_conversations.listeners

import com.twilio.conversations.Conversation
import com.twilio.conversations.ConversationListener
import com.twilio.conversations.Message
import com.twilio.conversations.Participant
import twilio.flutter.twilio_conversations.Mapper
import twilio.flutter.twilio_conversations.TwilioConversationsPlugin

class ConversationListener(private val conversationSid: String) : ConversationListener {
    override fun onMessageAdded(message: Message) {
        TwilioConversationsPlugin.debug("ConversationListener.onMessageAdded => messageSid = ${message.sid}")
        TwilioConversationsPlugin.flutterClientApi.messageAdded(
            conversationSid,
            Mapper.messageToPigeon(message)) {}
    }

    override fun onMessageUpdated(message: Message, reason: Message.UpdateReason) {
        TwilioConversationsPlugin.debug("ConversationListener.onMessageUpdated => messageSid = ${message.sid}, reason = $reason")
        TwilioConversationsPlugin.flutterClientApi.messageUpdated(
            conversationSid,
            Mapper.messageToPigeon(message),
            reason.toString()) {}
    }

    override fun onMessageDeleted(message: Message) {
        TwilioConversationsPlugin.debug("ConversationListener.onMessageDeleted => messageSid = ${message.sid}")
        TwilioConversationsPlugin.flutterClientApi.messageDeleted(
            conversationSid,
            Mapper.messageToPigeon(message)) {}
    }

    override fun onParticipantAdded(participant: Participant) {
        TwilioConversationsPlugin.debug("ConversationListener.onParticipantAdded => participantSid = ${participant.sid}")
        TwilioConversationsPlugin.flutterClientApi.participantAdded(
            conversationSid,
            Mapper.participantToPigeon(participant)) {}
    }

    override fun onParticipantUpdated(participant: Participant, reason: Participant.UpdateReason) {
        TwilioConversationsPlugin.debug("ConversationListener.onParticipantUpdated => participantSid = ${participant.sid}, reason = $reason")
        TwilioConversationsPlugin.flutterClientApi.participantUpdated(
            conversationSid,
            Mapper.participantToPigeon(participant),
            reason.toString()) {}
    }

    override fun onParticipantDeleted(participant: Participant) {
        TwilioConversationsPlugin.debug("ConversationListener.onParticipantDeleted => participantSid = ${participant.sid}")
        TwilioConversationsPlugin.flutterClientApi.participantDeleted(
            conversationSid,
            Mapper.participantToPigeon(participant)) {}
    }

    override fun onTypingStarted(conversation: Conversation, participant: Participant) {
        TwilioConversationsPlugin.debug("ConversationListener.onTypingStarted => conversationSid = ${conversation.sid}, participantSid = ${conversation.sid}")
        TwilioConversationsPlugin.flutterClientApi.typingStarted(
            conversationSid,
            Mapper.conversationToPigeon(conversation),
            Mapper.participantToPigeon(participant)) {}
    }

    override fun onTypingEnded(conversation: Conversation, participant: Participant) {
        TwilioConversationsPlugin.debug("ConversationListener.onTypingEnded => conversationSid = ${conversation.sid}, participantSid = ${participant.sid}")
        TwilioConversationsPlugin.flutterClientApi.typingEnded(
            conversationSid,
            Mapper.conversationToPigeon(conversation),
            Mapper.participantToPigeon(participant)) {}
    }

    override fun onSynchronizationChanged(conversation: Conversation) {
        TwilioConversationsPlugin.debug("ConversationListener.onSynchronizationChanged => sid: ${conversation.sid}, status: ${conversation.synchronizationStatus}")
        TwilioConversationsPlugin.flutterClientApi.synchronizationChanged(
            conversationSid,
            Mapper.conversationToPigeon(conversation)) {}
    }
}
